code/parachain/frame/composable-maths/src/dex/constant_product.rs

```diff
@@ -348,3 +350,86 @@ pub fn compute_deposit_lp(
 		Ok((overwritten_quote_amount, lp_to_mint))
 	}
 }
+
+/// Computes the decimal value of a a `PerThing` by taking the deconstructed parts of a `PerThing`
+/// and dividing them by `PerThing::one()`.
+///
+/// # Example
+/// ```rust,ignore
+/// let per_thing = PerMill::from_percent(10);
+/// assert_eq!(decimal_from_per_thing(per_thing), Decimal::new(10, 2));
+/// ```
+fn decimal_from_per_thing<T: PerThing>(per_thing: T) -> Result<Decimal, ArithmeticError> {
+	let numerator =
+		Decimal::from_u128(per_thing.deconstruct().into()).ok_or(ArithmeticError::Overflow)?;
+	let denominator =
+		Decimal::from_u128(T::one().deconstruct().into()).ok_or(ArithmeticError::Overflow)?;
+
+	numerator.safe_div(&denominator)
+}
+
+/// Computes the LP to mint on first deposit.
+///
+/// If `Ok`, returns a tuple containing `(lp_to_mint, fee)`.
+///
+/// Fees are currently always 0. Fees are normally charged to avoid fee-less swaps by adding and
+/// removing liquidity. With the initial deposit, these chances are so low that it is safe to
+/// process without a fee.
+///
+/// Paramaters:
+/// * `lp_token_details` - Vec of tuples containing `(token_deposit, token_balance, token_weight)`
+/// * `f` - Fee
+///
+/// https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#42-liquidity-provider-token-lpt-math-updates
+/// Equation 6
+fn compute_first_deposit_lp_<T: PerThing>(
+	pool_assets: Vec<(u128, u128, T)>,
+	_f: T,
+) -> Result<(u128, u128), ConstantProductAmmError> {
+	let k: u128 = pool_assets.len().try_into().map_err(|_| ArithmeticError::Overflow)?;
+	let product = pool_assets
+		.iter()
+		.try_fold(1, |product, (d_i, _b_i, _w_i)| product.safe_mul(d_i))?;
+
+	// NOTE: Zero fees on first deposit
+	Ok((k.safe_mul(&product)?, 0))
+}
+
+/// Computes the LP to mint on an existing deposit.
+///
+/// If `Ok`, returns a tuple containing `(lp_to_mint, fee)`.
+///
+/// Paramaters:
+/// * `p_supply` -
+/// * `d_k` - Deposit of token `k`
+/// * `b_k` - Balance of token `k`
+/// * `w_k` - Weight of token `k`
+/// * `f` - Fee
+///
+/// https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#42-liquidity-provider-token-lpt-math-updates
+/// Equation 5
+fn compute_deposit_lp_<T: PerThing>(
+	p_supply: u128,
+	d_k: u128,
+	b_k: u128,
+	w_k: T,
+	f: T,
+) -> Result<(u128, u128), ConstantProductAmmError> {
+	let p_supply = Decimal::from_u128(p_supply).ok_or(ArithmeticError::Overflow)?;
+	let d_k = Decimal::from_u128(d_k).ok_or(ArithmeticError::Overflow)?;
+	let b_k = Decimal::from_u128(b_k).ok_or(ArithmeticError::Overflow)?;
```

Comment:
Perhaps a helper method for this operation? The `RoundingDecimal` trait from the previous PR could be renamed to `DecimalExt` and then add these relevant methods there

Reply:
could also move `decimal_from_per_thing` there  if you're feeling funky, to be `Decimal::from_per_thing` instead

Reply:
897b5a30880567e776a15859cb64b37d703ceea5